apps/studio/src/components/Datatable/Datatable.tsx

```diff
@@ -0,0 +1,167 @@
+import type { LayoutProps, TableCellProps, TableProps } from "@chakra-ui/react"
+import type { Table as ReactTable } from "@tanstack/react-table"
+import {
+  Box,
+  Flex,
+  Spinner,
+  Table,
+  Tbody,
+  Td,
+  Text,
+  Th,
+  Thead,
+  Tr,
+  useMultiStyleConfig,
+} from "@chakra-ui/react"
+import { flexRender } from "@tanstack/react-table"
+
+import { DatatablePagination } from "./DatatablePagination"
+
+export interface DatatableProps<D> extends TableProps {
+  instance: ReactTable<D>
+  /**
+   * If provided, this number will be used for pagination instead of retrieving
+   * from react-table's filtered row count.
+   */
+  totalRowCount?: number
+  pagination?: boolean
+  /**
+   * If provided, this string will be used to display the total row count.
+   */
+  totalRowCountString?: string
+  isFetching?: boolean
+  tablePropOverrides?: Record<string, { td: TableCellProps }>
+  emptyPlaceholder?: React.ReactElement
+  overflow?: LayoutProps["overflow"]
+}
+
+export function createAccessor<T>(props: (keyof T)[]) {
+  return (row: T): string => {
+    return props.map((prop) => String(row[prop])).join(" ")
+  }
+}
+
+export const Datatable = <T extends object>({
+  instance,
+  isFetching,
+  pagination,
+  totalRowCount,
+  totalRowCountString,
+  tablePropOverrides,
+  emptyPlaceholder,
+  overflow = "auto",
+  ...tableProps
+}: DatatableProps<T>): JSX.Element => {
+  const { rows } = instance.getRowModel()
+  const styles = useMultiStyleConfig("Table", tableProps)
+
+  return (
+    <Flex flexDirection="column" layerStyle="shadow" pos="relative">
+      {isFetching && (
+        <>
+          <Flex
+            // white alpha to denote loading
+            bg="whiteAlpha.800"
+            bottom={0}
+            left={0}
+            p="1rem"
+            pos="absolute"
+            right={0}
+            top={0}
+            zIndex="1"
+          />
+          <Flex
+            bottom={0}
+            flex={1}
+            left={0}
+            pos="fixed"
+            right={0}
+            top={0}
+            w="100vw"
+            zIndex={2}
+          >
+            <Box m="auto">
+              <Spinner />
+            </Box>
+          </Flex>
+        </>
+      )}
```

Comment:
this component has to care about `loading` states but actually what i feel is the right abstraction here would be for the component here to only care about rendering the table and not care about loading. 

`Suspense` feels like a good fit here, we can omit hte `isLoading` and it'll still fit nicely! wdyt?

Reply:
cannot use suspense. We do not want the full fallback. We want to show the previous items in the table still, which suspense will not do (since it will swap out the current rendered table items with the suspense item completely). 
That is also the reason why data fetching is still done with `useQuery` and not `useSuspenseQuery`.

The loading state is a semi-transparent overlay.